docs/boring-bomb-architecture-doc.md

```diff
@@ -0,0 +1,115 @@
+    2. In the future if needed, we can add Auth functionality to the DB and write a frontend for Trust Assembly devs who don’t have access to the DB to add more `Datasources`, but in this MVP we’re making the decision to rely on direct DB inserts to add/update/delete `Datasources` and `DatasourceOwners`
+2. A cron job (or any function called periodically) will call a function on the Workers to fetch the `Datasource`s, and use them to insert jobs into the Worker Job Queue. Each job consists of a `Datasource` (for example a url like https://extelligence.substack.com/archive)
+3. The same Workers will subscribe to the job queue and one by one consume a job by using the `Datasources` to determine which URLs to fetch and parse. There will be some non-trivial logic here for how to use a `Datasource` along with existing `Articles` to determine which publications (i.e. specific article URL’s, like https://extelligence.substack.com/p/the-norwegian-butter-knife) need to have their headlines analyzed, and which already have `Article` s in the DB from a previous successfully completed job.
+4. Once the headline data is parsed, Workers will use OpenAI to analyze their original headlines & original article text to detect bias, and if the headline is biased then the Worker will use OpenAI to generate an unbiased headline.
+5. The worker will insert the unbiased headline into the DB as an `Article`. The Worker will also extract and insert the main article text into the `Article` . This isn’t useful for the MVP for Boring Bomb, but will be useful when we later move beyond the MVP to using a wiki to host main article text edits and Trust Assembly Adjudication.
+6. Finally, a user with the Trust Assembly browser extension visits a website, and the extension uses the API to fetch a headline for the article they’re reading. If no `Article` exists in our DB, then nothing will be changed about their website’s article. If one does exist, then the extension will use the unbiased headline from the `Article`
+
+## 🧑‍💼 Business Objects and Their Relations
+
+1. `Datasource`
+    1. id
+    2. parent_id (if the url to the same article has changed over time, then the new `Datasource` will have the prior `Datasource` as its parent_id)
+    3. base_url (i.e. [`https://extelligence.substack.com/archive`](https://extelligence.substack.com/archive))
+    4. owner_id
+    5. time_last_scraped
+2. `Article`
+    1. id
+    2. original_text
+    3. headline_text
+    4. edit_pairs (this will be an array of [original_text, edited text] tuples, where `original_text` is the original sentence/paragraph/section of text, and `edited_text` is the text to replace the `original_text` with)
```

Comment:
Rabbit-hole discussion, but worth pointing out alternate data models:
Another way to model this is with 3 tables:

Article:
- id
- original_text
- headline_text
- url
- datasource_id

ArticleVersion:
- id
- article_id
- previous_version_id
- text_diff
- headline_diff

ArticleEdit:
- id
- article_version_id
- user_id
- edit_object (JSON blob with edits at locations)

A new article created in our DB would create an Article and an ArticleVersion (with empty diffs). We could then build up the article contents with edits the same way git does

Reply:
OK, so when a user fetches an article, the process for arriving at the most recent headline would be:

1. fetch an article given a `url` along with all of its articleversions, ordered by `previous_version_id`
2. iterate through the articleversions, applying the `text_diff` and `headline_diffs` to the `original_text` and `headline_text`, respectively.
3. Finally, get the ArticleEdit corresponding to the latest article_id, and apply the edits from the `edit_object` to the article text we arrived at in step (2).

hmm I'm not seeing the main benefit over the earlier proposal, except that this makes very explicit how the article has changed over time (by storing those changes in `ArticleVersion`. This would be a useful property if we really cared about the changes, like if we were building a payments applications and every `ArticleVersion` was instead a `Transaction`, and we needed to keep track of the `Transactions` and use them to arrive at final Account balances. In this scenario it's really important to track the changes (i.e. the `text_diff` and `headline_diff` in `ArticleVersion`) explicitly. But in Trust Assembly's case I can't see why we care very much about the text changes; we really only care about the making sure we've got the most latest text of the Article so we can generate an accurate headline for it.

The second proposal also has some downsides, in that its:
1. more data to track in the DB
2. arriving at the final Article text and headline is slower and more complicated (thus more prone to bugs)

I prefer the first proposal, but please respond back with aspects of the second proposal I'm missing that make it preferable to the first.

Reply:
I propose storing the versions because any user edit is associated with a specific version. If I go to the wiki on 12/1 and edit some of the body text, then the article is changed on 12/2 in a location where the edit was made, we would want a way to account for this.

Also edit pairs with `original string`:`updated string` won't work because there's no way to reliably localize the edit. Let's say I'm a pedantic editor and I want to correct an `a` with a `the`. We can't just find all instances of `a` and replace with `the` -- we need a way to say where the edit was made. This would work where we associate edits with a version because there would be a list of changes that allow the edit to move to the updated location e.g. edit at position 20-30, article was updated to add 7 characters at position 10, edit shown in the new version is at position 27-37 (calculated at request time, no need to update the edit object itself).

This concern is well beyond MVP, but I don't want us to paint ourselves into a corner